pkg/apis/metal3/v1alpha1/baremetalhost_types.go

```diff
@@ -204,6 +207,66 @@ type BMCDetails struct {
 	DisableCertificateVerification bool `json:"disableCertificateVerification,omitempty"`
 }
 
+// HardwareRAIDVolume defines the desired configuration of volume in hardware RAID
+type HardwareRAIDVolume struct {
+	// Size (Integer) of the logical disk to be created in GiB.
+	// If unspecified or set be 0, the maximum capacity of disk will be used for logical disk.
+	// +kubebuilder:validation:Minimum=0
+	SizeGibibytes *int `json:"sizeGibibytes,omitempty"`
+
+	// RAID level for the logical disk. The following levels are supported: 0;1;1+0.
+	// +kubebuilder:validation:Enum="0";"1";"1+0"
+	Level string `json:"level" required:"true"`
+
+	// Name of the volume. Should be unique within the Node. If not specified, volume name will be auto-generated.
+	// +kubebuilder:validation:MaxLength=64
+	Name string `json:"name,omitempty"`
+
+	// Select disks with only rotational or solid-state storage
+	Rotational *bool `json:"rotational,omitempty"`
+
+	// Integer, number of disks to use for the logical disk. Defaults to minimum number of disks required
+	// for the particular RAID level.
+	// +kubebuilder:validation:Minimum=1
+	NumberOfPhysicalDisks *int `json:"numberOfPhysicalDisks,omitempty"`
+}
+
+// SoftwareRAIDVolume defines the desired configuration of volume in software RAID
+type SoftwareRAIDVolume struct {
+	// Size (Integer) of the logical disk to be created in GiB.
+	// If unspecified or set be 0, the maximum capacity of disk will be used for logical disk.
+	// +kubebuilder:validation:Minimum=0
+	SizeGibibytes *int `json:"sizeGibibytes,omitempty"`
+
+	// RAID level for the logical disk. The following levels are supported: 0;1;1+0.
+	// +kubebuilder:validation:Enum="0";"1";"1+0"
+	Level string `json:"level" required:"true"`
+
+	// Name of the volume. Should be unique within the Node. If not specified, volume name will be auto-generated.
+	// +kubebuilder:validation:MaxLength=64
+	Name string `json:"name,omitempty"`
+
+	// A list of device hints, the number of item must be greater than or equal to 2.
+	// +kubebuilder:validation:MinItems=2
+	PhysicalDisks []RootDeviceHints `json:"physicalDisks,omitempty"`
+}
+
+// RAIDConfig contains the configuration that are required to config RAID in Bare Metal server
+type RAIDConfig struct {
+	// The list of logical disks for hardware RAID, first volume is root volume
+	HardwareRAIDVolumes []HardwareRAIDVolume `json:"hardwareRAIDVolumes,omitempty"`
```

Comment:
I am sorry, i forget this comment. Could you give me a example in other programming languages?

Reply:
I am sorry, i forget this comment. Could you give me a example in other programming languages?

Reply:
I think we couldn't merge `hardware RAID` and `software RAID`, because the `PhysicalDisks` for `hardware RAID` and `software RAID` is different(or we can use `[]interface{}` type), and `hardware RAID` need `Name`/``, but `software RAID` needn't it. So i think it will provide some invalid options to users .

There is the description from [docs.openstack.org](https://docs.openstack.org/ironic/latest/admin/raid.html#backing-physical-disks):

> physical_disks - A list of physical disks to use as read by the RAID interface.
> For software RAID physical_disks is a list of device hints in the same format as used for Specifying the disk for deployment (root device hints). The number of provided hints must match the expected number of backing devices (repeat the same hint if necessary).

Example :

Hardware RAID:
```
{
  "logical_disks": [
    {
      "size_gb": 100,
      "raid_level": "5",
      "controller": "Smart Array P822 in Slot 3",
      "physical_disks": ["6I:1:5", "6I:1:6", "6I:1:7"],
      "is_root_volume": true
    }
  ]
}
```
Software RAID:
```
{
  "logical_disks": [
    {
      "size_gb": "MAX",
      "raid_level": "0",
      "controller": "software",
      "physical_disks": [
        {"size": "> 100"},
        {"size": "> 100"}
      ]
    }
  ]
}
```